src/flb_output_thread.c

```diff
@@ -333,7 +334,7 @@ static void output_thread(void *data)
         flb_sched_timer_cleanup(sched);
 
         /* Check if we should stop the event loop */
-        if (stopping == FLB_TRUE && mk_list_size(&th_ins->flush_list) == 0) {
+        if (stopping == FLB_TRUE && mk_list_size(&th_ins->flush_list) == 0 && mk_list_size(&th_ins->timer_coro_list) == 0) {
```

Comment:
why mk_list_size() for timer_coro_list in this case is not protected as in the others part of the code (mutex) ?

Reply:
The flush_list isn't protected either, I thought that in this control code, its guaranteed that this is the only code running in this thread and accessing the thread instance list... is that not true? Why is the flush_list not protected?

Reply:
I need to think on this some more.

Reply:
Matt is helping me realize that the problem is that the engine looks at these lists too, which means there need to be a lock that the engine acquires to ensure its allowed to safely review the list that the output thread owns

Reply:
>      * If the main engine (parent thread) needs to query the number of active
     * 'flushes' running by a threaded instance, then the access to the 'flush_list'
     * must be protected: we use 'flush_mutex for that purpose.

https://github.com/fluent/fluent-bit/blob/master/include/fluent-bit/flb_output_thread.h#L85

I htink I get it now. The engine needs to acquire the lock to read the list, and the output thread needs to acquire the lock for writes/deletes, so that the engine won't read it at the same time. If the output thread is just reading, then that's always safe. Assuming the output thread will modify the list.